### PR TITLE
[4.0] Convert plugin user profile to prepared statements

### PR DESCRIPTION
**Pull request by HLeithner**
# Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.

### Testing Instructions
Use the plugin in all ways you can think of.

Modify the user profile.

### Expected result
Nothing changed.

### CODE REVIEW

File: plugins/user/profile/field/tos.php [hunk 1]
```diff
@@ -86,9 +87,11 @@ protected function getLabel()
 
 			$db    = Factory::getDbo();
 			$query = $db->getQuery(true);
-			$query->select('id, alias, catid, language')
-				->from('#__content')
-				->where('id = ' . $tosArticle);
+
+			$query->select($db->quoteName(['id, alias, catid, language']))
```
Comment (SharkyKZ) [inline plugins/user/profile/field/tos.php]:
```suggestion
			$query->select($db->quoteName(['id', 'alias', 'catid', 'language']))
```

File: plugins/user/profile/field/tos.php [hunk 2]
```diff
@@ -86,9 +87,11 @@ protected function getLabel()
...
+				->from($db->quoteName('#__content'))
+				->where($db->quoteName('id') .' = :id')
```
Comment (SharkyKZ) [inline plugins/user/profile/field/tos.php]:
```suggestion
				->where($db->quoteName('id') . ' = :id')
```

File: plugins/user/profile/profile.php [hunk 3]
```diff
@@ -69,36 +70,32 @@ public function __construct(& $subject, $config)
...
 				$query = $db->getQuery(true)
-					->select(
-						array(
-							$db->quoteName('profile_key'),
-							$db->quoteName('profile_value'),
-						)
-					)
-					->from('#__user_profiles')
-					->where($db->quoteName('user_id') . ' = ' . $db->quote((int) $userId))
+					->select($db->quoteName(['profile_key','profile_value']))
```
Comment (Quy) [inline plugins/user/profile/profile.php]:
```suggestion
					->select($db->quoteName(['profile_key', 'profile_value']))
```

**Comment by nadjak77** [comment]
I have tested this item :white_check_mark: successfully on 347b89af4f4cd90a2d528ccf9a2696f8b00242e0<br /><br />works like descriped<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25544">issues.joomla.org/tracker/joomla-cms/25544</a>.</sub>

**Comment by JazParkyn** [comment]
I have tested this item :white_check_mark: successfully on 347b89af4f4cd90a2d528ccf9a2696f8b00242e0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25544">issues.joomla.org/tracker/joomla-cms/25544</a>.</sub>

**Comment by waader** [comment]
I have tested this item :red_circle: unsuccessfully on 347b89af4f4cd90a2d528ccf9a2696f8b00242e0<br /><br />Enable plugin "User - Profile" and in user profil set "Terms of Service" to required and assign an article to "Select TOS Article".

Then you should get the following error on the frontend registration form:
42703, 7, FEHLER: Spalte »id, alias, catid, language« existiert nicht LINE 1: SELECT "id, alias, catid, language" ^

I tested with postgresql.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25544">issues.joomla.org/tracker/joomla-cms/25544</a>.</sub>

File: plugins/user/profile/profile.php (hunk 3 above)
Comment (Quy) [inline plugins/user/profile/profile.php]: Revert

File: plugins/user/profile/profile.php [hunk 4]
```diff
@@ -428,20 +425,21 @@ public function onUserAfterSave($data, $isNew, $result, $error)
...
 				$key = 'profile.' . $key;
-				$key = $db->quote($key);
 			}
 
 			$query = $db->getQuery(true)
 				->delete($db->quoteName('#__user_profiles'))
-				->where($db->quoteName('user_id') . ' = ' . (int) $userId)
-				->where($db->quoteName('profile_key') . ' IN (' . implode(',', $keys) . ')');
+				->where($db->quoteName('user_id') . ' = :userid')
+				->whereIn($db->quoteName('profile_key'), $keys)
```
Comment (wilsonge) [inline plugins/user/profile/profile.php]: This needs to be type string

Comment (HLeithner) [reply plugins/user/profile/profile.php]: Fixed thanks

File: plugins/user/profile/profile.php (hunk 3 above)
Comment (HLeithner) [reply plugins/user/profile/profile.php]: done

File: plugins/user/profile/profile.php [hunk 5]
```diff
@@ -69,36 +70,33 @@ public function __construct(& $subject, $config)
...
-						array(
-							$db->quoteName('profile_key'),
-							$db->quoteName('profile_value'),
-						)
-					)
-					->from('#__user_profiles')
-					->where($db->quoteName('user_id') . ' = ' . $db->quote((int) $userId))
-					->where($db->quoteName('profile_key') . ' LIKE ' . $db->quote('profile.%'))
-					->order($db->quoteName('ordering'));
+										->select([
```
Comment (Quy) [inline plugins/user/profile/profile.php]: Decrease indentation.

File: plugins/user/profile/profile.php [hunk 6]
```diff
@@ -69,36 +79,33 @@ public function __construct(& $subject, $config)
 	public function onContentPrepareData($context, $data)
 	{
 		// Check we are manipulating a valid form.
-		if (!in_array($context, array('com_users.profile', 'com_users.user', 'com_users.registration', 'com_admin.profile')))
+		if (!in_array($context, ['com_users.profile', 'com_users.user', 'com_users.registration', 'com_admin.profile']))
 		{
 			return true;
 		}
-
-		if (is_object($data))
```
Comment (SharkyKZ) [inline plugins/user/profile/profile.php]: Was this removed intentionally?

Comment (HLeithner) [reply plugins/user/profile/profile.php]: Pretty sure this was a mistake, thanks

Comment (wilsonge) [reply plugins/user/profile/profile.php]: You need to push the change to fix it then ;)

Comment (HLeithner) [reply plugins/user/profile/profile.php]: done

File: plugins/user/profile/profile.php (hunk 5 above)
Comment (HLeithner) [reply plugins/user/profile/profile.php]: my IDE hates me...

File: plugins/user/profile/profile.php [hunk 7]
```diff
@@ -69,36 +86,37 @@ public function __construct(& $subject, $config)
...
 							$db->quoteName('profile_key'),
 							$db->quoteName('profile_value'),
-						)
+						]
 					)
-					->from('#__user_profiles')
-					->where($db->quoteName('user_id') . ' = ' . $db->quote((int) $userId))
+					->from($db->quoteName('#__user_profiles'))
+					->where($db->quoteName('user_id') . ' = :userid')
 					->where($db->quoteName('profile_key') . ' LIKE ' . $db->quote('profile.%'))
```
Comment (alikon) [inline plugins/user/profile/profile.php:111]: why isn't  binded ?
```suggestion
					->where($db->quoteName('profile_key') . ' LIKE  :profile'))
					->bind(':profile', 'profile.%')
```

File: plugins/user/profile/profile.php [hunk 8]
```diff
@@ -488,11 +523,12 @@ public function onUserAfterDelete($user, $success, $msg)
...
 		if ($userId)
 		{
-			$db = Factory::getDbo();
+			$db = $this->db;
 			$query = $db->getQuery(true)
 				->delete($db->quoteName('#__user_profiles'))
-				->where($db->quoteName('user_id') . ' = ' . $db->quote((int) $userId))
-				->where($db->quoteName('profile_key') . ' LIKE ' . $db->quote('profile.%'));
+				->where($db->quoteName('user_id') . ' = :userid')
+				->where($db->quoteName('profile_key') . ' LIKE ' . $db->quote('profile.%'))
```
Comment (alikon) [inline plugins/user/profile/profile.php:530]: same here

File: plugins/user/profile/profile.php (hunk 7 above)
Comment (HLeithner) [reply plugins/user/profile/profile.php:111]: because it's a constant and not a variable

File: plugins/user/profile/profile.php (hunk 8 above)
Comment (HLeithner) [reply plugins/user/profile/profile.php:530]: because it's a constant and not a variable

File: plugins/user/profile/profile.php (hunk 7 above)
Comment (alikon) [reply plugins/user/profile/profile.php:111]: oops 😃

File: plugins/user/profile/profile.php (hunk 8 above)
Comment (alikon) [reply plugins/user/profile/profile.php:530]: 🤕

Comment (alikon) [reply plugins/user/profile/profile.php:530]: I should stop reading code by phone

File: plugins/user/profile/profile.php [hunk 9]
```diff
@@ -69,36 +86,37 @@ public function __construct(& $subject, $config)
...
-		if (!in_array($context, array('com_users.profile', 'com_users.user', 'com_users.registration', 'com_admin.profile')))
+		if (!in_array($context, ['com_users.profile', 'com_users.user', 'com_users.registration', 'com_admin.profile']))
 		{
 			return true;
 		}
 
 		if (is_object($data))
 		{
-			$userId = $data->id ?? 0;
+			$userId = (int) $data->id ?? 0;
```
Comment (Quy) [inline plugins/user/profile/profile.php]:
```suggestion
			$userId = $data->id ?? 0;
```

> Notice: Undefined property: stdClass::$id in \plugins\user\profile\profile.php on line 96

Comment (HLeithner) [reply plugins/user/profile/profile.php]: Shouldn't it be ```$userId = (int) ($data->id ?? 0);```

Comment (Quy) [reply plugins/user/profile/profile.php]: `$data->id` is an integer value. It is up to you.

**Review by wilsonge — APPROVED** [review]

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 78e9bef97b40220624093a1e291f6cecf5345990<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25544">issues.joomla.org/tracker/joomla-cms/25544</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 78e9bef97b40220624093a1e291f6cecf5345990<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25544">issues.joomla.org/tracker/joomla-cms/25544</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25544">issues.joomla.org/tracker/joomla-cms/25544</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**